watertap/tools/oli_api/flash.py

```diff
@@ -0,0 +1,494 @@
+                    "name": name,
+                    "unit": conc[1],
+                    "value": conc[0],
+                    "charge": charge,
+                }
+            )
+
+        for k, v in self.water_analysis_properties.items():
+            if v["value"] is not None:
+                if isinstance(v["value"], list):
+                    self.water_analysis_properties[k]["value"] = v["value"][0]
+                if all(k != i["name"] for i in self.water_analysis_input_list):
+                    self.water_analysis_input_list.append(v)
+
+        input_list = deepcopy(self.water_analysis_input_list)
+        return input_list
+
+    def _oli_units(self, conc, unit):
+        """
+        Converts concentrations between specified units.
+
+        :param conc: concentration value for a solute
+        :param unit: concentration unit for a solute
+
+        :return converted_value: tuple with converted concentration and OLI unit string
+        """
+        to_units = input_unit_set["molecularConcentration"]
+        converted_value = (
+            pyunits.convert_value(conc, unit, to_units["pyomo_unit"]),
+            to_units["oli_unit"],
+        )
+        return converted_value
+
+    def _set_prescaling_calculation_mode(self, use_scaling_rigorous):
+        """
+        Sets prescaling computation method based on argument.
+
+        :param use_scaling_rigorous: boolean indicating desired state of 'rigorous' and 'estimated' optional properties
+        """
+
+        if bool(use_scaling_rigorous) == bool(
+            self.optional_properties["prescalingTendenciesRigorous"]
+        ):
+            return
+        new_values = {
+            k: not v for k, v in self.optional_properties.items() if "prescaling" in k
+        }
+        self.optional_properties.update(new_values)
+
+    def build_flash_calculation_input(
+        self,
+        flash_method,
+        state_vars,
+        water_analysis_output=None,
+        inflows_phase="total",
+        use_scaling_rigorous=True,
+    ):
+        """
+        Builds a base dictionary required for OLI flash analysis.
+
+        :param flash_method: string name of OLI flash flash_method to use
+        :param state_vars: dictionary containing solutes, temperatures, pressure, and units
+        :param water_analysis_output: dictionary to extract inflows from (required if not wateranalysis flash)
+        :param inflows_phase: string indicating desired phase of inflows
+        :param use_scaling_rigorous: boolean switch to use estimated or rigorous solving for prescaling metrics
+
+        :return inputs: dictionary containing inputs for specified OLI flash analysis
+        """
+
+        inputs = {"params": {}}
+
+        if flash_method == "wateranalysis":
+            inputs["params"] = {
+                "waterAnalysisInputs": self._build_input_list(state_vars)
+            }
+        else:
+            if not water_analysis_output:
```

Comment:
Hmm-- so this is basically saying-- if you want to you use isothermal flash, the only way to do that is to run water analysis first, which takes true species as input, and in the output from that, you can grab apparent species, which is the required input for isothermal flash.

However, what if I just want to provide apparent species from the get-go? Although in our team's usage of OLI thus far, we've used water analysis followed by isothermal flash, it's possible that the general user might want to just start off with isothermal flash. So, I wonder if you considered including something in `_build_input_list` that allow you to build inputs based on true species (water analysis) OR apparent species (isothermal flash). 

In that sense, the user wouldn't necessarily be restricted to having to run a water analysis first if they simply wanted to begin with apparent species when initiating a model. Does that make sense?

Reply:
BTW- I don't think this point should hold up this PR necessarily, unless you see a quick way to do this and a burning desire to do so. I mostly just want to get your thoughts on this, hear whether there some subtle complexities with taking that approach, and then decide if we should create an issue for this to address at a later point.

Reply:
My impression was that Isothermal Flash required true species as input, so I forced that part of the workflow. If that's not the case, the workflow could certainly be made more flexible to allow for different inputs

Reply:
Ah, so Isothermal Flash takes in apparent species, while we can provide true species to Water Analysis. Water Analysis gives apparent species in its ouput though, and that's why we used that output as input to Isothermal Flash. The main reason we wanted to do that was because OLI makes certain optional properties accessible through Isothermal Flash results that are NOT produced by Water Analysis (or weren't at the time at least).

So to clarify with a simple example, using water analysis, we would feed in Na_+ and Cl_-. Using isothermal flash, we would use NaCl as input.

Reply:
Ok, I'll look into that